subsys/net/lib/sockets/sockets.c

```diff
@@ -173,6 +175,76 @@ int z_impl_zsock_socket(int family, int type, int proto)
 	return -1;
 }
 
+int z_impl_zsock_socketpair(int family, int type, int proto, int sv[2]) {
+
+	int res;
+
+	/* Use a temporary socket vector since sv[] should not be modified
+         * on failure. See POSIX.1-2016 */
+	int temp_sv[2];
+
+	/* Zero-initializing this (with the exception of setting the family)
+	 * has the same effect as using INADDR_ANY_INIT or IN6ADDR_ANY_INIT.
```

Comment:
> Zero-initializing this (with the exception of setting the family) has the same effect as using INADDR_ANY_INIT or IN6ADDR_ANY_INIT.

Perhaps. But why do you bind to INADDR_ANY? Please provide normative reference that this is required, or let me conjecture that this is serious security issue, where you open your "inter-thread communication" (per https://github.com/zephyrproject-rtos/zephyr/issues/24366) to attacks from the wide Internet.

Actually, you unlikely to provide such a reference, because normative spec says essebtially the opposite:

> The socketpair() function shall create an **unbound** pair of connected sockets

(emphasis mine). So, let me give -1 to this for now, and let's think how to deal with that.

Reply:
Hmm... Interesting. Normally when I look up this man page, I see that it says "unnamed" rather than "unbound". E.g. in Linux, it only works with AF_UNIX.

Also, I just realized I'm missing a listen(2) call, which would normally be necessary for e.g. AF_INET / SOCK_STREAM so it likely would not work in those cases and it would also expose a race condition / security issue, unless there was a way, as you say, to connect two sockets without binding.

Reply:
I would actually prefer to only support AF_UNIX, but since there is no such support in Zephyr yet, I thought I should work with what is available.

Reply:
> I would actually prefer to only support AF_UNIX

That would be natural course of thinking, and then socketpair() implementation wouldn't be part of network subsystem per se.

As you know, native Zephyr way to communicate data streams between threads are Zephyr pipes. If I were to implement this in the most useful way, and yet bound the scope-creep, I'd consider following way:

1. See if socketpair(AF_UNIX) can be reasonably implemented in terms of classical pipe().
2. If so, implement pipe() for POSIX subsys (I don't think we have it), and then socketpair() in top of it.

Feel free to give this idea a thought.

Reply:
@pfalcon - if you would prefer as well, I could remove AF_INET and focus on adding AF_UNIX support instead.

Reply:
> I could remove AF_INET and focus on adding AF_UNIX support instead. Its use could be limited to socketpair() at least initially

+1 to this

Reply:
> > I would actually prefer to only support AF_UNIX
> 
> That would be natural course of thinking, and then socketpair() implementation wouldn't be part of network subsystem per se.

true...

> As you know, native Zephyr way to communicate data streams between threads are Zephyr pipes. If I were to implement this in the most useful way, and yet bound the scope-creep, I'd consider following way:
> 
> 1. See if socketpair(AF_UNIX) can be reasonably implemented in terms of classical pipe().
> 2. If so, implement pipe() for POSIX subsys (I don't think we have it), and then socketpair() in top of it.
> 
> Feel free to give this idea a thought.

Good idea. I wasn't actually sure what would be the best way to implement it in Zephyr. Are there additional / special markers that need to be inserted in order to support SOCK_DGRAM?

Reply:
> Good idea. I wasn't actually sure what would be the best way to implement it in Zephyr.

@cfriedt, before discussing this further, I'd like to get your reply to https://github.com/zephyrproject-rtos/zephyr/issues/24366#issuecomment-613965163 . Because again, if you want to just signal an event to poll(), then eventfd() is a better choice.

(OTOH, if you really want to support SOCK_DGRAM (in addition to natural choice of SOCK_STREAM), then might be then reusing existing network socket infrastructure actually makes sense after all (or it would be indeed a bunch of code duplication between network socket subsystem and non-network socket impl).)

Reply:
@pfalcon - I would be fine just supporting SOCK_STREAM (SOCK_RAW is even probably unnecessary and I don't even know if there is a difference to SOCK_STREAM in this case, but some people prefer it for AF_UNIX).

Reply:
> > Zero-initializing this (with the exception of setting the family) has the same effect as using INADDR_ANY_INIT or IN6ADDR_ANY_INIT.
> 
> Perhaps. But why do you bind to INADDR_ANY? Please provide normative reference that this 

Originally, I was just thinking INADDR_LOOPBACK, but then saw that `lo` requires an additional CONFIG option, but yea.. let's just forget about AF_INET.